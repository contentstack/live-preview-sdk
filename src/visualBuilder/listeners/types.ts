import { VisualEditorCslpEventDetails } from "../types/visualBuilder.types";

interface EventListenerHandlerParams {
    event: MouseEvent;
    overlayWrapper: HTMLDivElement | null;
    visualEditorContainer: HTMLDivElement | null;
    previousSelectedEditableDOM: Element | null;
    focusedToolbar: HTMLDivElement | null;
    resizeObserver: ResizeObserver;
    eventDetails: VisualEditorCslpEventDetails;
    customCursor: HTMLDivElement | null;
}

export default EventListenerHandlerParams;
