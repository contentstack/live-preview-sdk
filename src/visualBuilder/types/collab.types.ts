//User
export interface IUserDTO {
    display?: string;
    email: string;
    uid: string;
    firstName?: string | null;
    lastName?: string | null;
}

export declare interface IUserState {
    mentionsList: Array<IMentionList>;
    userMap: IUserMap;
    currentUser: IUserDTO;
}

export declare type IUserMap = { [key: string]: IUserDTO };

// Invite Metadata and Context
export declare interface IInviteMetadata {
    users: Array<IUserDTO>;
    currentUser: IUserDTO;
    inviteUid: string;
}

// Mention Related Types
export declare type IMentionItem = { id: string; display: string };

export declare interface IMentionList {
    display: string;
    email?: string;
    uid?: string;
}

export declare type IMentionedList = Array<IMentionItem>;

// Comment State
export declare interface ICommentState {
    message: string;
    toUsers?: Array<IMentionItem>;
    images?: string[];
    createdBy: string;
    author: string;
}

// Message DTO
export declare interface IMessageDTO {
    _id: string;
    threadUid: string;
    message: string;
    author: string;
    toUsers?: string[];
    images?: string[];
    createdAt: string;
    updatedAt?: string;
    createdBy: string;
}

// Thread Context
export declare interface IThreadContext {
    inviteMetadata: IInviteMetadata;
    commentCount: number;
    error: IErrorState;
    userState: IUserState;
    onCreateComment: (data: ICommentPayload) => Promise<ICommentResponse>;
    onEditComment: (data: IEditCommentArgs) => Promise<ICommentResponse>;
    onDeleteComment: (data: IDeleteCommentArgs) => Promise<IDefaultAPIResponse>;
    onDeleteThread: (data: IDeleteThreadArgs) => void;
    setThreadState: (
        state:
            | IThreadPopupState
            | ((prevState: IThreadPopupState) => IThreadPopupState)
    ) => void;
    onClose: (isResolved?: boolean) => void;
    setError: Function;
    editComment: string;
    activeThread: IActiveThread;
    setActiveThread: (thread: IActiveThread) => void;
    createNewThread: () => Promise<any>;
}

export interface IThreadPopupState {
    commentCount: number;
    userState: IUserState;
    isLoading: boolean;
    comments: Array<IMessageDTO>;
    editComment: string;
}

export interface IThreadRenderStatus {
    threadId: string;
    attempts: number;
    isRendered: boolean;
}

export interface ICommentDTO {
    _id: string;
    threadUid: string;
    message: string;
    author: string;
    toUsers: string[];
    images: string[];
    createdAt: string;
    createdBy: string;
    updatedAt?: string;
}

export interface IDefaultAPIResponse {
    notice: string;
}

export interface ICommentResponse extends IDefaultAPIResponse {
    comment: ICommentDTO;
}

export interface IFetchCommentsResponse {
    comments: Array<ICommentDTO>;
    count: number;
}

export interface ICommentPayload {
    threadUid: string;
    commentPayload: ICommentState;
}
export interface IThreadPayload {
    elementXPath: string;
    position: {
        x: number;
        y: number;
    };
    author: string;
    pageRoute: string;
    inviteUid: string;
    createdBy: string;
}

export interface IThreadDTO {
    _id: string;
    author: string;
    inviteUid: string;
    position: {
        x: number;
        y: number;
    };
    elementXPath: string;
    isElementPresent: boolean;
    pageRoute: string;
    createdBy: string;
    sequenceNumber: number;
    threadState: number;
    createdAt: string;
    updatedAt?: string;
}
export interface IThreadResponseDTO extends IDefaultAPIResponse {
    thread: IThreadDTO;
}

export interface IEditCommentArgs {
    threadUid: string;
    commentUid: string;
    payload: ICommentState;
}

export interface IDeleteCommentArgs {
    threadUid: string;
    commentUid: string;
}

export interface IDeleteThreadArgs {
    threadUid: string;
}

export interface IThreadResolveArgs {
    threadUid: string;
    payload: {
        threadState: number;
    };
}

export interface MissingThreadsInfo {
    payload: {
        isElementPresent: boolean;
    };
    threadUids: string[];
}

export interface IErrorState {
    hasError: boolean;
    message: string;
}

export interface IActiveThread extends Partial<IThreadDTO> {
    _id: string;
}

export interface IFetchComments {
    threadUid: string;
    offset: number;
    limit: number;
}

export interface toggleCollabPopupEvent {
    threadUid: string;
    action: string;
}

export interface ICommentActionBar {
    mode: "edit" | "view";
    commentUser: IUserDTO;
    currentUser: IUserDTO;
    commentUID?: string | undefined;
}

export interface ICommentCard {
    comment: IMessageDTO | null;
    onClose: (isResolved?: boolean) => void;
    userState: IUserState;
    mode: "edit" | "view";
    handleOnSaveRef: React.MutableRefObject<any>;
}

export interface ICommentResolvedText {
    comment: IMessageDTO;
    userState: IUserState;
}

export interface ICommentTextAreaProps {
    userState: IUserState;
    onClose: (isResolved?: boolean) => void;
    handleOnSaveRef: React.MutableRefObject<any>;
    comment?: IMessageDTO | null;
}

export interface IThreadPopup {
    onCreateComment: (data: ICommentPayload) => Promise<ICommentResponse>;
    onEditComment: (data: IEditCommentArgs) => Promise<ICommentResponse>;
    onDeleteComment: (data: IDeleteCommentArgs) => Promise<IDefaultAPIResponse>;
    onDeleteThread: (data: IDeleteThreadArgs) => void;
    onClose: (isResolved?: boolean) => void;
    onResolve: (data: IThreadResolveArgs) => Promise<IThreadResponseDTO>;
    inviteMetadata: IInviteMetadata;
    loadMoreMessages: (data: IFetchComments) => Promise<IFetchCommentsResponse>;
    activeThread: IActiveThread;
    setActiveThread: (thread: IActiveThread) => void;
    createNewThread: () => Promise<IThreadResponseDTO>;
}

export interface IThreadBody {
    handleOnSaveRef: React.MutableRefObject<any>;
    onClose: (isResolved?: boolean) => void;
    userState: IUserState;
    isLoading: boolean;
    comments: IMessageDTO[];
    fetchingMore: boolean;
    editComment: string;
}

export interface IThreadFooter {
    onClose: (isResolved?: boolean) => void;
    handleOnSaveRef: React.MutableRefObject<any>;
    isDisabled: boolean;
    editComment: string;
}

export interface IThreadHeader {
    onClose: (isResolved?: boolean) => void;
    displayResolve: boolean;
    onResolve: (data: IThreadResolveArgs) => Promise<IDefaultAPIResponse>;
    commentCount: number;
    activeThread: IActiveThread;
}

export interface ICollabConfig {
    collab: {
        fromShare: boolean;
        pauseFeedback: boolean;
        enable: boolean;
        isFeedbackMode: boolean;
        inviteMetadata: IInviteMetadata;
        payload: IThreadDTO[];
    };
}

export interface IThreadIdentifier {
    threadUid: string;
}

export interface IThreadRemove {
    threadUids: string[];
    updateConfig?: boolean;
}

export interface IThreadReopen {
    thread: IThreadDTO;
}
