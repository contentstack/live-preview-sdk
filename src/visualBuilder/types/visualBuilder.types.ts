import { CslpData } from "../../cslp/types/cslp.types";

export interface VisualEditorCslpEventDetails {
    editableElement: Element;
    cslpData: string;
    fieldMetadata: CslpData;
}
