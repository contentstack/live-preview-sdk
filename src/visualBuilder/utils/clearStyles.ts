const originalStyles = new WeakMap<
    HTMLElement,
    {
        visibility?: string;
        transition?: string;
        animation?: string;
    }
>();

export function clearVisibelityStyles(element: HTMLElement) {
    const originalStyleValues = {
        visibility: element.style.visibility,
        transition: element.style.transition,
        animation: element.style.animation,
    };

    originalStyles.set(element, originalStyleValues);

    element.style.visibility = "hidden";
    element.style.transition = "none";
    element.style.animation = "none";
}

export function restoreVisibelityStyles(element: HTMLElement) {
    const storedStyles = originalStyles.get(element);

    if (storedStyles) {
        element.style.visibility = storedStyles.visibility || "";
        element.style.transition = storedStyles.transition || "";
        element.style.animation = storedStyles.animation || "";

        originalStyles.delete(element);
    }
}
