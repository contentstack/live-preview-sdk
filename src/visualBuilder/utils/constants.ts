import { FieldDataType } from "./types/index.types";

export const numericInputRegex = /^-?\d*(\.\d*)?([eE][-+]?\d*)?$/;

export const VISUAL_BUILDER_FIELD_TYPE_ATTRIBUTE_KEY = "data-cslp-field-type";

export const VISUAL_BUILDER_CHANNEL_ID = "visual-builder";

export const LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX = 2;

// Minimum distance from top edge to prevent toolbar from being hidden
export const TOP_EDGE_BUFFER = 42;
export const RIGHT_EDGE_BUFFER = 180;

export const TOOLBAR_EDGE_BUFFER = 8;

export const DATA_CSLP_ATTR_SELECTOR = "data-cslp";

/**
 * The field that can be directly modified using contenteditable=true.
 * This includes all text fields like title and numbers.
 */
export const ALLOWED_INLINE_EDITABLE_FIELD: FieldDataType[] = [
    FieldDataType.SINGLELINE,
    FieldDataType.MULTILINE,
    FieldDataType.NUMBER,
];

export const ALLOWED_MODAL_EDITABLE_FIELD: FieldDataType[] = [
    FieldDataType.HTML_RTE,
    FieldDataType.MARKDOWN_RTE,
    FieldDataType.JSON_RTE,
    FieldDataType.CUSTOM_FIELD,
    FieldDataType.LINK,
    FieldDataType.ISODATE,
    FieldDataType.URL,
];

export const ALLOWED_REPLACE_FIELDS: FieldDataType[] = [
    FieldDataType.REFERENCE,
    FieldDataType.FILE,
];

export const DEFAULT_MULTIPLE_FIELDS: FieldDataType[] = [
    FieldDataType.GLOBAL_FIELD,
    FieldDataType.GROUP,
    FieldDataType.BLOCK,
];

export const mentionLimit = 20;
export const maxMessageLength = 500;
