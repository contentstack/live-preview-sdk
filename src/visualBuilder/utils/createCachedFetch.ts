/**
 * Creates a cached async fetch function with support for any number of arguments
 * @param fetchFn - The async function to cache
 * @param uidResolver - Function that generates a unique ID from the arguments passed to fetchFn
 * @returns A cached version of the fetch function with the same signature
 */
export function createCachedFetch<TArgs extends any[], TResult>(
    fetchFn: (...args: TArgs) => Promise<TResult>,
    uidResolver: (...args: TArgs) => string = (...args) => JSON.stringify(args)
): {
    (...args: TArgs): Promise<TResult>;
    clearCache: () => void;
} {
    // Cache storage
    const cache = new Map<string, TResult>();
    // Track in-flight requests
    const pendingPromises = new Map<string, Promise<TResult>>();

    /**
     * The cached fetch function
     * @param args - Arguments to pass to the original fetch function
     * @returns Promise that resolves with the data
     */
    async function cachedFetch(...args: TArgs): Promise<TResult> {
        // Generate unique ID for these arguments
        const uid = uidResolver(...args);

        // Return cached value if available
        if (cache.has(uid)) {
            return cache.get(uid)!;
        }

        // Return existing promise if request is already in progress
        if (pendingPromises.has(uid)) {
            return pendingPromises.get(uid)!;
        }

        // Create new promise for this request
        const promise = fetchFn(...args)
            .then((data) => {
                // Store result in cache
                cache.set(uid, data);
                // Remove from pending
                pendingPromises.delete(uid);
                return data;
            })
            .catch((error) => {
                // Clean up on error
                pendingPromises.delete(uid);
                throw error;
            });

        // Store the promise
        pendingPromises.set(uid, promise);
        return promise;
    }

    // Add clearCache method to the function
    cachedFetch.clearCache = () => {
        cache.clear();
        pendingPromises.clear();
    };

    return cachedFetch;
}
