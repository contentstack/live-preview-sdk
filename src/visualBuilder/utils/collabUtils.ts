import {
    ICommentState,
    IMentionedList,
    IMentionItem,
    IMessageDTO,
    IUserDTO,
    IUserState,
} from "../types/collab.types";
import { maxMessageLength, mentionLimit } from "./constants";
import { uniqBy } from "lodash-es";
import DOMPurify from "dompurify";

const escapeRegExp = (string: string): string => {
    return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
};

/**
 * Generates the title for the thread based on the number of comments.
 * @param {number} commentCount - The number of comments.
 * @returns {string} The title for the thread.
 */
export const getThreadTitle = (commentCount: number): string => {
    if (commentCount === 0) return "Add New Feedback";
    return commentCount === 1 ? "1 Feedback" : `${commentCount} Feedbacks`;
};

/**
 * returns the available email.
 * @param {IUserDTO} user - The user object.
 * @returns {string} The user's email.
 */
export const getUserName = (user: IUserDTO): string => {
    return user.email;
};

/**
 * Validates the comment length and the number of mentions.
 * @param {string} comment - The comment message.
 * @param {IMentionedList} toUsers - The list of mentioned users.
 * @returns {string} The error message if validation fails, otherwise an empty string.
 */
export const validateCommentAndMentions = (
    comment: string,
    toUsers: IMentionedList
): string => {
    if (comment.length > maxMessageLength) {
        return `Limit exceeded. You can have a maximum length of ${maxMessageLength} characters.`;
    }
    if (toUsers.length > mentionLimit) {
        return `Limit exceeded. You can tag a maximum of ${mentionLimit} users.`;
    }
    return "";
};

/**
 * Removes mentions that no longer exist in the message.
 * @param {string} message - The comment message.
 * @param {IMentionedList} toUsers - The list of mentioned users.
 * @returns {Object} The updated lists of mentioned users.
 */
export const filterOutInvalidMentions = (
    message: string,
    toUsers: IMentionedList
) => {
    const to_users_temp = toUsers.filter((user) =>
        message.includes(user.display)
    );

    return {
        toUsers: uniqBy(to_users_temp, "id"),
    };
};

/**
 * Replaces mention placeholders with display names in the comment message.
 * @param {IMessageDTO | undefined} comment - The comment object.
 * @param {IUserState} userState - The user state containing user and role maps.
 * @param {"text" | "html"} profile - The format for the output message, either plain text or HTML.
 * @returns {string | undefined} The formatted message or undefined if the comment is not provided.
 */
export const getMessageWithDisplayName = (
    comment: IMessageDTO | undefined | null,
    userState: IUserState,
    profile: "text" | "html"
): string | undefined => {
    if (!comment) return undefined;

    let tempText = comment.message;

    comment?.toUsers?.forEach((user) => {
        const userPattern = new RegExp(`{{${user}}}`, "g");
        const userData = userState.userMap[user];
        const replacement =
            profile === "html"
                ? `<b>@${userData.display || getUserName(userData)}</b>`
                : `@${userData.display || getUserName(userData)}`;
        tempText = tempText.replace(userPattern, replacement);
    });

    return tempText;
};

/**
 * Sanitizes HTML content to prevent XSS attacks.
 * @param {any} dirty - The unsanitized HTML content.
 * @returns {string} The sanitized HTML content.
 */
export const sanitizeData = (dirty: any): string => {
    return DOMPurify.sanitize(dirty, { USE_PROFILES: { html: true } });
};

/**
 * Constructs the comment body with mentions replaced by their unique identifiers.
 * @param {ICommentState} state - The state containing the comment and mentions.
 * @returns {Object} The comment body containing the sanitized message and mentioned users.
 */
export const getCommentBody = (state: ICommentState): ICommentState => {
    let finalMessage = state.message.trim().replace(/\s+/g, " ");
    const comment = {
        message: finalMessage,
        toUsers: [],
        images: [],
        createdBy: state.createdBy,
        author: state.author,
    };

    const updateMentionToUID = (
        entity: IMentionItem,
        result: Array<string>
    ) => {
        const displayName = entity.display;

        const escapedDisplayName = escapeRegExp(displayName);
        const regexUser = new RegExp(escapedDisplayName, "g");
        finalMessage = finalMessage.replace(regexUser, `{{${entity.id}}}`);
        result.push(entity.display);
    };

    state.toUsers?.forEach((user) => updateMentionToUID(user, comment.toUsers));

    comment.message = finalMessage;
    return comment;
};
