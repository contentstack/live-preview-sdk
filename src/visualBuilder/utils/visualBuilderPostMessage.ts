import { EventManager } from "@contentstack/advanced-post-message";
import { getCommunicationTarget } from "../../common";
import { VISUAL_BUILDER_CHANNEL_ID } from "./constants";

let visualBuilderPostMessage: EventManager | undefined;

if (typeof window !== "undefined") {
    const target = getCommunicationTarget();
    
    if (target) {
        visualBuilderPostMessage = new EventManager(VISUAL_BUILDER_CHANNEL_ID, {
            target,
            debug: false,
            // suppressErrors: true,
        });
    }
}

export default visualBuilderPostMessage;
