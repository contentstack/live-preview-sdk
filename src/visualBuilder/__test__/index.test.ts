import crypto from "crypto";
import { getFieldSchemaMap } from "../../__test__/data/fieldSchemaMap";
import { getElementBytestId, mockGetBoundingClientRect, sleep, triggerAndWaitForClickAction, waitForBuilderSDKToBeInitialized, waitForToolbaxToBeVisible } from "../../__test__/utils";
import Config from "../../configManager/configManager";
import { FieldSchemaMap } from "../utils/fieldSchemaMap";
import { waitFor, screen, cleanup } from "@testing-library/preact";
import { VisualBuilderPostMessageEvents } from "../utils/types/postMessage.types";
import { VisualBuilder } from "../index";
import visualBuilderPostMessage from "../utils/visualBuilderPostMessage";
import { Mock } from "vitest";

const INLINE_EDITABLE_FIELD_VALUE = "Hello World";

vi.mock("../utils/visualBuilderPostMessage", async () => {
    const { getAllContentTypes } = await vi.importActual<
        typeof import("../../__test__/data/contentType")
    >("../../__test__/data/contentType");
    const contentTypes = getAllContentTypes();

    return {
        __esModule: true,
        default: {
            send: vi.fn().mockImplementation((eventName: string) => {
                if (eventName === "init")
                    return Promise.resolve({
                        contentTypes,
                    });
                return Promise.resolve();
            }),
            on: vi.fn(),
        },
    };
});

Object.defineProperty(globalThis, "crypto", {
    value: {
        getRandomValues: (arr: Array<any>) => crypto.randomBytes(arr.length),
    },
});
// Increase the timeout for the test
describe("Visual builder", () => {
    beforeAll(() => {
        FieldSchemaMap.setFieldSchema(
            "all_fields",
            getFieldSchemaMap().all_fields
        );
        Config.set("mode", 2);
        vi.spyOn(
            document.documentElement,
            "clientWidth",
            "get"
        ).mockReturnValue(100);
        vi.spyOn(
            document.documentElement,
            "clientHeight",
            "get"
        ).mockReturnValue(100);
        vi.spyOn(document.body, "scrollHeight", "get").mockReturnValue(100);
    });

    beforeEach(() => {
        (visualBuilderPostMessage?.send as Mock).mockClear();
        document.getElementsByTagName("html")[0].innerHTML = "";
        cleanup();
    });

    afterAll(() => {
        FieldSchemaMap.clear();
    });

    test("should append a visual builder container to the DOM", async () => {
        let visualBuilderDOM = document.querySelector(
            ".visual-builder__container"
        );

        expect(visualBuilderDOM).toBeNull();

        const x = new VisualBuilder();
        await waitForBuilderSDKToBeInitialized(visualBuilderPostMessage);

        visualBuilderDOM = document.querySelector(
            `[data-testid="visual-builder__container"]`
        );

        expect(document.querySelector('[data-testid="visual-builder__cursor"]')).toBeInTheDocument();
        expect(document.querySelector('[data-testid="visual-builder__focused-toolbar"]')).toBeInTheDocument();
        expect(document.querySelector('[data-testid="visual-builder__hover-outline"]')).toBeInTheDocument();
        expect(document.querySelector('[data-testid="visual-builder__overlay__wrapper"]')).toBeInTheDocument();
        x.destroy();
    }, { timeout: 20 * 1000 });

    test("should add overlay to DOM when clicked", async () => {
        const h1Tag = document.createElement("h1");
        h1Tag.textContent = INLINE_EDITABLE_FIELD_VALUE;
        h1Tag.setAttribute(
            "data-cslp",
            "all_fields.blt58a50b4cebae75c5.en-us.modular_blocks.0.block.single_line"
        );
        document.body.appendChild(h1Tag);
        mockGetBoundingClientRect(h1Tag);
        const x = new VisualBuilder();
        await triggerAndWaitForClickAction(visualBuilderPostMessage, h1Tag);
        await waitFor(() => {
            const overlayOutline = document.querySelector('[data-testid="visual-builder__overlay--outline"]');
            expect(overlayOutline).toHaveStyle({
                top: "10px",
                left: "10px",
                width: "10px",
                height: "5px",
                "outline-color": "rgb(113, 92, 221)"
            });
        })
        x.destroy();
    }, { timeout: 20 * 1000 });

    describe("on click, the sdk", () => {
        afterEach(() => {
            document.getElementsByTagName("html")[0].innerHTML = "";
        });

        test("should do nothing if data-cslp not available", async () => {
            const h1 = document.createElement("h1");

            document.body.appendChild(h1);
            const x = new VisualBuilder();
            await triggerAndWaitForClickAction(visualBuilderPostMessage, h1, {skipWaitForFieldType: true});

            expect(h1).not.toHaveAttribute("contenteditable");
            expect(h1).not.toHaveAttribute("data-cslp-field-type");
            x.destroy();
        });

        describe("inline elements must be contenteditable", () => {
            let visualBuilder: VisualBuilder;
            let h1: HTMLHeadingElement
            beforeAll(() => {
                (visualBuilderPostMessage?.send as Mock).mockImplementation(
                    (eventName: string, args) => {
                        if (
                            eventName ===
                            VisualBuilderPostMessageEvents.GET_FIELD_DATA
                        ) {
                            const values: Record<string, any> = {
                                single_line: INLINE_EDITABLE_FIELD_VALUE,
                                multi_line: INLINE_EDITABLE_FIELD_VALUE,
                                file: {
                                    uid: "fileUid",
                                },
                            };
                            return Promise.resolve({
                                fieldData: values[args.entryPath],
                            });
                        } else if (
                            eventName ===
                            VisualBuilderPostMessageEvents.GET_FIELD_DISPLAY_NAMES
                        ) {
                            const names: Record<string, string> = {
                                "all_fields.blt58a50b4cebae75c5.en-us.single_line":
                                    "Single Line",
                                "all_fields.blt58a50b4cebae75c5.en-us.multi_line":
                                    "Multi Line",
                                "all_fields.blt58a50b4cebae75c5.en-us.file":
                                    "File",
                            };
                            return Promise.resolve({
                                [args.cslp]: names[args.cslp],
                            });
                        }
                        return Promise.resolve({});
                    }
                );
            });

            beforeEach(async () => {
                document.getElementsByTagName("html")[0].innerHTML = "";
                h1 = document.createElement("h1");
                h1.textContent = INLINE_EDITABLE_FIELD_VALUE;
                mockGetBoundingClientRect(h1);
                h1.setAttribute(
                    "data-cslp",
                    "all_fields.blt58a50b4cebae75c5.en-us.single_line"
                );

                document.body.appendChild(h1);
                visualBuilder = new VisualBuilder();
            })
            afterEach(() => {
                visualBuilder.destroy();
            })
            test("single line should be contenteditable", async () => {
                await triggerAndWaitForClickAction(visualBuilderPostMessage, h1);

                await waitFor(() => {
                    expect(h1).toHaveAttribute("contenteditable");
                    expect(h1).toHaveAttribute("data-cslp-field-type", "singleline");
                })
            }, { timeout: 20 * 1000 });

            test("multi line should be contenteditable", async () => {
                h1.setAttribute(
                    "data-cslp",
                    "all_fields.blt58a50b4cebae75c5.en-us.multi_line"
                );
                await triggerAndWaitForClickAction(visualBuilderPostMessage, h1);

                await waitFor(() => {
                    expect(h1).toHaveAttribute("contenteditable");
                    expect(h1).toHaveAttribute("data-cslp-field-type", "multiline");
                })
            }, { timeout: 20 * 1000 });

        });
    });
}, { timeout: 20 * 1000 });
