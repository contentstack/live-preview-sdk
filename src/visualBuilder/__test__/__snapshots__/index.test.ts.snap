// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Visual builder > on click, the sdk > inline elements must be contenteditable > file should render a replacer and remove when it is not 1`] = `
<h1
  data-cslp="all_fields.blt58a50b4cebae75c5.en-us.file"
  data-cslp-field-type="file"
/>
`;

exports[`Visual builder > on click, the sdk > inline elements must be contenteditable > file should render a replacer and remove when it is not 2`] = `
<button
  class="visual-builder__replace-button visual-builder__button visual-builder__button--secondary go2828613387 go2545267607 go3805905658"
  data-testid="visual-builder-replace-file"
  data-tooltip="Replace"
>
  <svg
    data-testid="visual-builder__replace-asset-icon"
    fill="none"
    height="16"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M13.8903 3.99632L11.5039 1.60984C11.3574 1.46339 11.12 1.46339 10.9735 1.60984C10.8271 1.75628 10.8271 1.99372 10.9735 2.14017L12.7198 3.88649H2.87506C2.39181 3.88649 2.00006 4.27824 2.00006 4.76149V6.13114C2.00006 6.33825 2.16795 6.50614 2.37506 6.50614C2.58217 6.50614 2.75006 6.33825 2.75006 6.13114V4.76149C2.75006 4.69245 2.80603 4.63649 2.87506 4.63649H12.7198L10.9735 6.38281C10.8271 6.52925 10.8271 6.76669 10.9735 6.91314C11.12 7.05958 11.3574 7.05958 11.5039 6.91314L13.8903 4.52665C14.0368 4.3802 14.0368 4.14277 13.8903 3.99632Z"
      fill="#475161"
    />
    <path
      d="M2.10983 12.0054L4.49632 14.3919C4.64277 14.5383 4.8802 14.5383 5.02665 14.3919C5.1731 14.2455 5.1731 14.008 5.02665 13.8616L3.28033 12.1152H13.1251C13.6084 12.1152 14.0001 11.7235 14.0001 11.2402V9.87059C14.0001 9.66348 13.8322 9.49559 13.6251 9.49559C13.418 9.49559 13.2501 9.66348 13.2501 9.87059V11.2402C13.2501 11.3093 13.1942 11.3652 13.1251 11.3652H3.28033L5.02665 9.61893C5.1731 9.47248 5.1731 9.23504 5.02665 9.0886C4.8802 8.94215 4.64277 8.94215 4.49632 9.0886L2.10983 11.4751C1.96339 11.6215 1.96339 11.859 2.10983 12.0054Z"
      fill="#475161"
    />
  </svg>
</button>
`;

exports[`Visual builder > on click, the sdk > inline elements must be contenteditable > multi line should be contenteditable 1`] = `
<h1
  contenteditable="true"
  data-cslp="all_fields.blt58a50b4cebae75c5.en-us.multi_line"
  data-cslp-field-type="multiline"
>
  Hello World
</h1>
`;

exports[`Visual builder > on click, the sdk > inline elements must be contenteditable > single line should be contenteditable 1`] = `
<h1
  contenteditable="true"
  data-cslp="all_fields.blt58a50b4cebae75c5.en-us.single_line"
  data-cslp-field-type="singleline"
>
  Hello World
</h1>
`;

exports[`Visual builder > on click, the sdk > should do nothing if data-cslp not available 1`] = `
<body>
  <h1 />
  <div
    class="go263407482 visual-builder__container"
    data-testid="visual-builder__container"
  >
    <style>
      
       @import url("https://fonts.googleapis.com/css2?family=Inter:ital,opsz,wght@0,14..32,100..900;1,14..32,100..900&display=swap");

       [data-cslp] [contenteditable="true"] {
            outline: none;
        }
        
        @keyframes visual-builder__spinner {
            0% {
                transform: rotate(0deg);
            }
            100% {
                transform: rotate(360deg);
            }
        }


    </style>
    <div
      class="go1688149084 visual-builder__cursor"
      data-testid="visual-builder__cursor"
    />
    <div
      class="go3860994383 visual-builder__overlay__wrapper"
      data-testid="visual-builder__overlay__wrapper"
    >
      <div
        class="visual-builder__overlay visual-builder__overlay--top go2262268276"
        data-testid="visual-builder__overlay--top"
      />
      <div
        class="visual-builder__overlay visual-builder__overlay--left go2262268276"
        data-testid="visual-builder__overlay--left"
      />
      <div
        class="visual-builder__overlay visual-builder__overlay--right go2262268276"
        data-testid="visual-builder__overlay--right"
      />
      <div
        class="visual-builder__overlay visual-builder__overlay--bottom go2262268276"
        data-testid="visual-builder__overlay--bottom"
      />
      <div
        class="visual-builder__overlay--outline go3391008271"
        data-testid="visual-builder__overlay--outline"
      />
    </div>
    <div
      class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go217287799 go1719569523"
      data-testid="visual-builder__hover-outline"
    />
    <div
      class="visual-builder__focused-toolbar go104737281"
      data-testid="visual-builder__focused-toolbar"
    />
  </div>
</body>
`;

exports[`Visual builder > should append a visual builder container to the DOM 1`] = `
<div
  class="go263407482 visual-builder__container"
  data-testid="visual-builder__container"
>
  <style>
    
       @import url("https://fonts.googleapis.com/css2?family=Inter:ital,opsz,wght@0,14..32,100..900;1,14..32,100..900&display=swap");

       [data-cslp] [contenteditable="true"] {
            outline: none;
        }
        
        @keyframes visual-builder__spinner {
            0% {
                transform: rotate(0deg);
            }
            100% {
                transform: rotate(360deg);
            }
        }


  </style>
  <div
    class="go1688149084 visual-builder__cursor"
    data-testid="visual-builder__cursor"
  />
  <div
    class="go3860994383 visual-builder__overlay__wrapper"
    data-testid="visual-builder__overlay__wrapper"
  >
    <div
      class="visual-builder__overlay visual-builder__overlay--top go2262268276"
      data-testid="visual-builder__overlay--top"
    />
    <div
      class="visual-builder__overlay visual-builder__overlay--left go2262268276"
      data-testid="visual-builder__overlay--left"
    />
    <div
      class="visual-builder__overlay visual-builder__overlay--right go2262268276"
      data-testid="visual-builder__overlay--right"
    />
    <div
      class="visual-builder__overlay visual-builder__overlay--bottom go2262268276"
      data-testid="visual-builder__overlay--bottom"
    />
    <div
      class="visual-builder__overlay--outline go3391008271"
      data-testid="visual-builder__overlay--outline"
    />
  </div>
  <div
    class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go217287799 go1719569523"
    data-testid="visual-builder__hover-outline"
  />
  <div
    class="visual-builder__focused-toolbar go104737281"
    data-testid="visual-builder__focused-toolbar"
  />
</div>
`;

exports[`visual builder DOM > should have an overlay over the element 1`] = `
<div
  class="go842373233 visual-builder__overlay__wrapper"
  data-testid="visual-builder__overlay__wrapper"
>
  <div
    class="visual-builder__overlay visual-builder__overlay--top go2262268276"
    data-testid="visual-builder__overlay--top"
  />
  <div
    class="visual-builder__overlay visual-builder__overlay--left go2262268276"
    data-testid="visual-builder__overlay--left"
  />
  <div
    class="visual-builder__overlay visual-builder__overlay--right go2262268276"
    data-testid="visual-builder__overlay--right"
  />
  <div
    class="visual-builder__overlay visual-builder__overlay--bottom go2262268276"
    data-testid="visual-builder__overlay--bottom"
  />
  <div
    class="visual-builder__overlay--outline go3391008271"
    data-testid="visual-builder__overlay--outline"
  />
</div>
`;

exports[`visual builder DOM > should have an overlay over the element 2`] = `
<div
  class="go842373233 visual-builder__overlay__wrapper visible"
  data-testid="visual-builder__overlay__wrapper"
>
  <div
    class="visual-builder__overlay visual-builder__overlay--top go2262268276"
    data-testid="visual-builder__overlay--top"
    style="top: 0px; left: 0px; width: 100%; height: calc(10px);"
  />
  <div
    class="visual-builder__overlay visual-builder__overlay--left go2262268276"
    data-testid="visual-builder__overlay--left"
    style="left: 0px; top: 10px; height: 5px; width: 10px;"
  />
  <div
    class="visual-builder__overlay visual-builder__overlay--right go2262268276"
    data-testid="visual-builder__overlay--right"
    style="left: 20px; top: 10px; height: 5px; width: 80px;"
  />
  <div
    class="visual-builder__overlay visual-builder__overlay--bottom go2262268276"
    data-testid="visual-builder__overlay--bottom"
    style="top: 20px; height: 80px; left: 0px; width: 100%;"
  />
  <div
    class="visual-builder__overlay--outline go3391008271"
    data-testid="visual-builder__overlay--outline"
    style="top: 10px; height: 5px; width: 10px; left: 10px; outline-color: rgb(113, 92, 221);"
  />
</div>
`;
